File: Deepspace/lib/SpaceStation.rb
#encoding:utf-8

require_relative "SpaceStationToUI"
require_relative "SuppliesPackage"

module Deepspace

#
# @author Miguel Ángel Fernández Gutiérrez, Sergio Quijano Rey
class SpaceStation

	# Class atributes
	# ==========================================================================

	# @!attribute [Integer] Maximum fuel quantity that a space station can have
	@@MAXFUEL = 100

	# @!attribute [Float] Shield units lost per each shot unit taken
	@SHIELDLOSSPERUNITSHOT = 0.1

	# Constructors
	# ==========================================================================
	
	# Class initializer
	# @param _name [String] name of the space station
	# @param _supplies [SuppliesPackage] starting fuel units, weapons and shields
	def initialize(_name, _supplies)
		# !@attribute [String] name of the space station
		@name = _name

		# !@attribute [Float] Parametrizes ammunition power
		@ammoPower = 0.0

		# !@attribute [Float] Parametrizes fuel units
		@fuelUnits = 0.0
		
		# !@attribute [Integer] Number of medals
		@nMedals = 0
		
		# !@attribute [Float] Parametrizes shield power
		@shieldPower = 0.0
		
		# !@attribute [Damage] Pending damage
		@pendingDamage = nil

		# !@attribute [Array<Weapon>] Array of weapons
		@weapons = []

		# !@attribute [Array<ShieldBooster] Array of shield boosters
		@shieldBoosters = []
		
		# !@attribute [Hangar] Hangar
		@hangar = nil
		
		# Supplies are added
		receiveSupplies(_supplies)
	end

	# Getters
	# ==========================================================================

	attr_reader :ammoPower, :fuelUnits, :hangar, :name, :nMedals,
			:pendingDamage, :shieldBoosters, :shieldPower, :weapons

	# Gets the speed of the space station.
	# Speed is calculated as fraction of fuel units and max fuel possible
	# @param [Float] percentage of speed, that's to say, a number in [0, 1]
	def speed
		if @@MAXFUEL == 0
			raise "WARNING, zero division at SpaceStation.speed()"
			return 0
		else
			return @fuelUnits.to_f / @@MAXFUEL
		end
	end

	# Checks the state of the space ship.
	# Valid state means no pending damage or pending damage with no effect
	# @return [Boolean] true, if the space ship is on a valid state;
	#                   false, otherwise
	def validState
		if @pendingDamage.nil?
			return false
		else
			#TEST return @pendingDamage.length == 0 || @pendingDamage.hasNoEffect
			return @pendingDamage.hasNoEffect
		end
	end

	# Description
	# 	Gets UI representation of the object
	# Returns
	# 	SpaceStationToUI, the UI representation
	def getUIVersion
		if validState
			return SpaceStationToUI.new(self)
		else
			puts "WARNING! Not UI version for space station on invalid state"
			return nil
		end
	end

	def to_s
		out = "[Space Station]-> Name: #{@name}\n"
		out += "\tNo. Medals: #{@nMedals}, Fuel units: #{@fuelUnits.round(2)}, ammoPower: #{@ammoPower}, shieldPower: #{@shieldPower}\n"
		out += "\tWeapons: [#{@weapons.join(' ,')}]\n"
		out += "\tShieldBoosters: [#{@shieldBoosters.join(', ')}]\n"
		out += "\tHangar: #{@hangar}\n"
		out += "\tPendingDamage: #{@pendingDamage}\n"
		out += "-------- end of Space Station >> #{@name} << --------"
		return out
	end
	
	# Setters
	# ==========================================================================

	# Assigns the fuel of the space station
	# @param f [Float] fuel value to be assigned
	def assignFuelValue(f)
		if f < @@MAXFUEL
			@fuelUnits = f
		else
			@fuelUnits = @@MAXFUEL
		end
	end

	# Adjusts certain damage to some weapon and shieldBoosters lists, and its
	# value is stored in the object
	# @param d [Damage] the damage to be set
	def setPendingDamage(d)
		@pendingDamage = d.adjust(@weapons, @shieldBoosters)
	end

	# If pending damage has no effect, fixes the atribute to nil
	def cleanPendingDamage
		if !@pendingDamage.nil?
			if @pendingDamage.hasNoEffect
				@pendingDamage = nil
			end
		end
	end

	# Tries to add a weapon to the hangar
	# @param [Weapon] the weapon to add
	# @return [Boolean] true, if weapon is successfully added;
	#                   false, if the operation fails
	def receiveWeapon(w)
		if hangar != nil
			return @hangar.addWeapon(w)
		else
			return false
		end
	end

	# Tries to add a shield booster to the hangar
	# @param [ShieldBooster] the shield booster to add
	# @return [Boolean] true, if booster is successfully added;
	#                   false, if the operation fails
	def receiveShieldBooster(s)
		if hangar != nil
			return @hangar.addShieldBooster(s)
		else
			return false
		end
	end

	# Tries to add a hangar.
	# If there's already a hangar, this method has no effect
	# @param h [Hangar] the hangar to add
	def receiveHangar(h)
		if @hangar.nil?
			@hangar = Hangar.newCopy(h) # Security copy
		end
	end

	# Discards current hangar (nil reference)
	def discardHangar
		@hangar = nil
	end

	# If a hangar is available, discards a weapon from it, in a certain
	# position
	# @param i [Integer] index where the weapon that wants to be discarded is
	#                    located in the hangar
	def discardWeaponInHangar(i)
		if !@hangar.nil?
			@hangar.removeWeapon(i)
		else
			raise "WARNING! Trying to discard a weapon where no hangar is available, at SpaceStation.discardWeaponInHangar()"
		end
	end

	# If a hangar is available, discards a shield booster from it, in a certain
	# position
	# @param i [Integer] index where the booster that wants to be discarded is
	#                    located in the hangar
	def discardShieldBoosterInHangar(i)
		if !@hangar.nil?
			@hangar.removeShieldBooster(i)
		else
			raise "WARNING! Trying to discard a shield booster where no hangar is available, at SpaceStation.discardShieldBoosterInHangar()"
		end
	end

	# Shot, shield and fuel power increase by a certain supplies package
	# @param s [SuppliesPackage] the supplies to add
	def receiveSupplies(s)
		@ammoPower += s.ammoPower
		assignFuelValue(@fuelUnits+s.fuelUnits)
		@shieldPower += s.shieldPower
	end

	# A weapon from the hangar is mounted to be used.
	# If method runs successfully, weapon is erased from Hangar, and the weapon
	# is added to the collection of weapons in use
	# @param i [Integer] index of the weapon to mount
	def mountWeapon(i)
		if @hangar.nil?
			raise "WARNING! No hangar available at SpaceStation.mountWeapon()"
		else
			# New weapon is deleted from the hangar
			new_weapon = @hangar.removeWeapon(i) 
			if !new_weapon.nil?
				@weapons << new_weapon
			else
				raise "WARNING! Trying to add nil weapon on SpaceStation.mountWeapon()"
			end
		end
	end

	# A shield booster from the hangar is mounted to be used.
	# If method runs successfully, booster is erased from Hangar, and the weapon
	# is added to the collection of boosters in use
	# @param i [Integer] index of the booster to mount
	def mountShieldBooster(i)
		if @hangar.nil?
			raise "WARNING! No hangar available, at SpaceStation.mountShieldBooster()"
		else
			new_shield = @hangar.removeShieldBooster(i)
			if new_shield.nil?
				raise "WARNING! Trying to add nil shield on Spacestation.mountShieldBooster()"
			else
				@shieldBoosters << new_shield
			end
		end
	end

	# The spaceships moves. Therefore, fuel units decrease
	def move
		@fuelUnits -= @fuelUnits*speed
		if @fuelUnits <= 0
			@fuelUnits = 0
		end
	end

	# Deletes all mounted weapons and mounted shields with no uses left
	def cleanUpMountedItems
		# Filtering weapons
		@weapons = @weapons.select{|weapon|  weapon.uses > 0}

		# Filtering shields
		@shieldBoosters = @shieldBoosters.select{|shield| shield.uses > 0}
	end

	# WIP -- Práctica 3
	# Description:
	# Returns:
	# 	Float
	def fire
	end
	
	# Description:
	# Returns:
	# 	Float
	def protection
	end

	# WIP -- Práctica 3
	# Description:
	# Parameters:
	# 	shot: Float
	# Returns:
	# 	ShotResult
	def receiveShot(shot)
	end

	# WIP -- Práctica 3
	# Description
	# Parameters:
	# 	loot: Loot
	def setLoot(loot)
	end

	# WIP -- Práctica 3
	# Description:
	# Parameters:
	# 	i: Integer
	def discardWeapon(i)
	end

	# WIP -- Práctica 3
	# Description:
	# Parameters:
	# 	i: Integer
	def discardShieldBooster(i)
	end

	# Private Specifiers
	#=======================================================================
	private :assignFuelValue, :cleanPendingDamage
end

end	# Deepspace

File: deepspace/src/deepspace/SpaceStation.java
package deepspace;

import java.util.ArrayList;
import java.util.stream.Collectors; // For filtering an ArrayList

/**
 * Class to represent a Space Station
 *
 * @author Miguel Angel Fernandez
 * @author Sergio Quijano Rey
 * */
class SpaceStation implements SpaceFighter{
    // Class Attributes
    //==========================================================================
    private static final int MAXFUEL = 100;                         //< The max ammount of fuel that the space station can have
    private static final float SHIELDLOSSPERUNITSHOT = (float)0.1;  //< The shield loss when an unit of shot power is received

    // Instance Attributes
    //==========================================================================
    private float ammoPower;                            //< The ammo power of the space station
    private float fuelUnits;                            //< The fuel units left
    private String name;                                //< The name of the space station
    private int nMedals;                                //< The number of medals that the space station have
    private float shieldPower;                          //< The power of the space station's shield
    private Damage pendingDamage;                       //< The pending damage done to the space station
    private ArrayList<Weapon> weapons;                  //< The mounted weapons of the space station
    private ArrayList<ShieldBooster> shieldBoosters;    //< The mounted shieldBoosters of the space station
    private Hangar hangar;                              //< The hangar which holds unmounted shields and weapons
    
    // Constructors
    //==========================================================================
    /**
     * @brief The constructor of the class
     * @param n, the name of the space station
     * @param supplies, the starting supplies package with ammoPower, shieldPower and fuelUnits
     * */
    SpaceStation(String n, SuppliesPackage supplies){
        // The name given as a parameter
        name = n;

        // Default values for all the attributes
        ammoPower = (float)0.0;
        fuelUnits = (float)0.0;
        nMedals = 0;
        shieldPower = (float)0.0;
        pendingDamage = null;
        weapons = new ArrayList<>(0);
        shieldBoosters = new ArrayList<>(0);
        hangar = null;

        // We add the supplies given as a parameter -- WIP overridable method call on constructor, may lead to bugs
        receiveSupplies(supplies);
    }

    /**
     * Copy contructor
     * @param other, the object to copy
     * */
    SpaceStation(SpaceStation other){
        // This elements do not need to be copied securely
        ammoPower = other.ammoPower;
        fuelUnits = other.fuelUnits;
        name = other.name;
        nMedals = other.nMedals;
        shieldPower = other.shieldPower;

        // This elements need a secure copy
        if(other.pendingDamage != null){
            pendingDamage = other.pendingDamage.copy();
        }
        if(other.weapons != null){
            weapons = new ArrayList<>(other.weapons);
        }
        if(other.shieldBoosters != null){
            shieldBoosters = new ArrayList<>(other.shieldBoosters);
        }
        if(other.hangar != null){
            hangar = new Hangar(other.hangar);
        }       
    }

    // Getters
    //==========================================================================
    /**
     * @brief ammoPower getter
     * */
    public float getAmmoPower(){
        return ammoPower;
    }

    /**
     * @brief fuelUnits getter
     * */
    public float getFuelUnits(){
        return fuelUnits;
    }

    /**
     * @brief hangar getter
     * */
    public Hangar getHangar(){
        return hangar;
    }

    /**
     * @brief name getter
     * */
    public String getName(){
        return name;
    }

    /**
     * @brief nMedals getter
     * */
    public int getNMedals(){
        return nMedals;
    }

    /**
     * @brief pendingDamage getter
     * */
    public Damage getPendingDamage(){
        return pendingDamage;
    }

    /**
     * @brief shieldBoosters getter
     * */
    public ArrayList<ShieldBooster> getShieldBoosters(){
        // Copy return
        return new ArrayList<>(shieldBoosters);
    }

    /**
     * @brief weapons getter
     * */
    public ArrayList<Weapon> getWeapons(){
        // Copỳ return
        return new ArrayList<>(weapons);
    }

    /**
     * @brief shieldPower getter
     * */
    public float getShieldPower(){
        return shieldPower;
    }

    /**
     * @brief returns the speed of the space station
     *        This speed is the fraction between fuelUnits and MAXFUEL
     *        Therefore, speed is in [0, 1] interval
     * */
    public float getSpeed(){
        return fuelUnits / MAXFUEL;
    }

    /**
     * @brief Checks wether or not the space station is on a valid state
     *        Valid state means no pending damage or pending damage with no effect
     * @return true, if the space station is on a valid state
     *         false, if the space station is not on a valid state
     * */
    public boolean validState(){
        if(pendingDamage == null){
            return true;
        }else{
            return pendingDamage.hasNoEffect();
        }

    }

    /**
     * @brief Gets the UI version
     * @return the UI Version
     * */
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }

    /**
     * @brief gets the string representation of the object
     * @return the string representation
     * */
    @Override
    public String toString(){
        return  "SpaceStation(\n" + 
                "\tammoPower = " + ammoPower + "\n" +
                "\tfuelUnits = " + fuelUnits + "\n" +
                "\tname = " + name + "\n" +
                "\tnMedals = " + nMedals + "\n" +
                "\tshieldPower = " + shieldPower + "\n" +
                "\tpendingDamage = " + pendingDamage + "\n" +
                "\tweapons = " + weapons + "\n" +
                "\tshieldBoosters = " + shieldBoosters + "\n" +
                "\tHangar = " + hangar + "\n" +
                ")";
    }

    // Setters
    //==========================================================================
    /**
     * The space station receives a loot
     * The card dealer gives us elements as specified on the loot
     * The elements are stored with their respective methods
     * @param loot, the loot which is received
     * @return the transformation that the loot indicates (may be NOTRANSFORM)
     *
     * BUG -- no funciona como deberiBUG
     * */
    public Transformation setLoot(Loot loot){
        // We get a card dealer for the elements received at the loot
        CardDealer dealer = CardDealer.getInstance();

        // We receive the hangars from the loot
        for(int i = 0; i < loot.getNHangars(); i++){
            Hangar local_hangar = dealer.nextHangar();
            receiveHangar(local_hangar); 
        }

        // We receive the supplies
        for(int i = 0; i < loot.getNSupplies(); i++){
            SuppliesPackage supplies = dealer.nextSuppliesPackage();
            receiveSupplies(supplies); 
        }

        // We receive the weapons
        for(int i = 0; i < loot.getNWeapons(); i++){
            Weapon weapon = dealer.nextWeapon();
            if(receiveWeapon(weapon) == false){
                System.out.println("Warning! Weapon could not be added propperly on SpaceStation.setLoot()");
            }
        }

        // We receive the shieldBoosters
        for(int i = 0; i < loot.getNShields(); i++){
            ShieldBooster sh = dealer.nextShieldBooster();
            if(receiveShieldBooster(sh) == false){
                System.out.println("Warning! ShieldBooster could not be added propperly on SpaceStation.setLoot()");
            }
        }

        // We receive the medals
        nMedals = nMedals + loot.getNMedals();

        // We return the transformation
        if(loot.getEfficient()){
            return Transformation.GETEFFICIENT;
        }else if(loot.spaceCity()){
            return Transformation.SPACECITY;
        }else{
            return Transformation.NOTRANSFORM;
        }
    }

    /**
     * @brief the adjusted damage is calculated and stored
     * @param d, the damage to be adjusted to the list of weapons and shields of the hangar
     *
     * WIP -- Necesita ser testeado intensamente para ver que no da fallos
     * */
    public void setPendingDamage(Damage d){
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    // Receive Elements
    //==========================================================================

    /**
     * @brief Receives an hangar
     *        If there is already an hangar this method has no effect
     * @param h, the hangar to receive
     * */
    public void receiveHangar(Hangar h){
        if(hangar == null){
            hangar = h;
        }
    }

    /**
     * @brief Receives a supplies package
     *        The ammoPower, fuelUnits and shieldPower is incremented by the
     *        SuppliesPackage
     * @param s, the supplies to receive
     * */
    public void receiveSupplies(SuppliesPackage s){
        // Ammo power is incremented
        ammoPower = ammoPower + s.getAmmoPower();

        // Shield power is incremented
        shieldPower = shieldPower + s.getShieldPower();

        // Fuel units are incremented
        assignFuelValue(fuelUnits + s.getFuelUnits());
    }

    /**
     * @brief Tries to add a shieldBooster to the hangar
     * @param s, the shieldBooster to add
     * @return  true, if the shieldBooster is added successfully
     *          false, if the shieldbooster could not be added
     * */
    public boolean receiveShieldBooster(ShieldBooster s){
        // Security check
        if(hangar == null){
            System.out.println("WARNING! No hangar on SpaceStation.receiveShieldBooster()");
            return false;
        }

        // We try to add the shield
        return hangar.addShieldBooster(s);
    }

    /**
     * @brief Tries to add a weapon to the hangar
     * @param w, the weapon to add
     * @return  true, if the weapon is added successfully
     *          false, if the weapon could not be added
     * */
    public boolean receiveWeapon(Weapon w){
        // Security check
        if(hangar == null){
            System.out.println("WARNING! No hangar on SpaceStation.receiveWeapon()");
            return false;
        }

        // We try to add the weapon
        return hangar.addWeapon(w);
    }

    // Mount Elements
    //==========================================================================
    
    /**
     * @brief Tries to mount a shieldBooster from the hangar
     * @param i, the position of the shieldBooster in the hangar
     * */
    public void mountShieldBooster(int i){
        if(hangar == null){
            System.out.println("WARNING! No hangar on SpaceStation.mountShieldBooster()");
        }else{
            // The weapon is removed from the hangar
            ShieldBooster s = hangar.removeShieldBooster(i);

            // The weapon is added to the list of weapons if it is not null
            if(s == null){
                System.out.println("WARNING! No shield on SpaceStation.mountShieldBooster()");
            }else{
                shieldBoosters.add(s);
            }
        }
        
    }

    /**
     * @brief Tries to mount a weapon from the hangar
     * @param i, the position of the weapon in the hangar
     * */
    public void mountWeapon(int i){
        if(hangar == null){
            System.out.println("WARNING! No hangar on SpaceStation.mountWeapon()");
        }else{
            // The weapon is removed from the hangar
            Weapon w = hangar.removeWeapon(i);

            // The weapon is added to the list of weapons if it is not null
            if(w == null){
                System.out.println("WARNING! No weapon on SpaceStation.mountWeapon()");
            }else{
                weapons.add(w);
            }
        }
    }

    // Discard Elements
    //==========================================================================
    /**
     * @brief Removes the mounted weapons and shieldBoosters with no uses left
     * */
    public void cleanUpMountedItems(){
        // Cleaning weapons -- WIP programacion funcional
        weapons = new ArrayList<>(weapons.stream().filter(weapon -> weapon.getUses() > 0).collect(Collectors.toList()));

        // Cleaning shieldBoosters -- WIP programacion funcional
        shieldBoosters = new ArrayList<>(shieldBoosters.stream().filter(shield -> shield.getUses() > 0).collect(Collectors.toList()));
    }

    /**
     * @brief Discards an hangar, that's to say, it is set to null
     * */
    public void discardHangar(){
        hangar = null;
    }

    /**
     * @brief Discards a shield in use, due to a pending damage
     *        Therefore, the pending damage is actualized
     * @param i, the position of the shield to discard
     * */
    public void discardShieldBooster(int i){
        if(i >= 0 && i < shieldBoosters.size()){
            ShieldBooster s = shieldBoosters.remove(i); // A shield is removed

            if(pendingDamage != null){
                pendingDamage.discardShieldBooster();   // The pending damage has to be actualized
                cleanPendingDamage();
            }

        }else{
            System.out.println("WARNING! Bad given index on SpaceStation.discardShieldBooster()");
        }
    }

    /**
     * @brief Discards a ShieldBooster from the hangar
     * @param i, the position of the shield booster in the hangar
     * */
    public void discardShieldBoosterInHangar(int i){
        if(hangar == null){
            System.out.println("WARNING! No hangar on SpaceStation.discardShieldBoosterInHangar()");
        }else{
            hangar.removeShieldBooster(i);
        }
    }

    /**
     * @brief A weapon in use is discarded, due to a pendingDamage
     *        Therefore, the pending damage is actualized
     * @param i, the position of the weapon to discard
     * */
    public void discardWeapon(int i){
        if(i >= 0 && i < weapons.size()){
            Weapon w = weapons.remove(i);   // A weapon is removed

            if(pendingDamage != null){
                pendingDamage.discardWeapon(w); // The pending damage has to be actualized
                cleanPendingDamage();
            }

        }else{
            System.out.println("WARNING! Bad given index on SpaceStation.discardWeapon()");
        }
    }

    /**
     * @brief Discards a weapon from the hangar
     * @param i, the position of the weapon in the hangar
     * */
    public void discardWeaponInHangar(int i){
        if(hangar == null){
            System.out.println("WARNING! No hangar on SpaceStation.discardWeaponInHangar()");
        }else{
            hangar.removeWeapon(i);
        }
    }

    // Space Station actions methods
    //==========================================================================
    /**
     * @brief Makes a shot and returns its power
     * @return the power of the shot made
     * */
    @Override
    public float fire(){
        float factor = (float) 1.0; // This variable is used for acummulating the extra power given by weapons 

        // We use the mounted weapons
        for(Weapon w : weapons){
            factor = factor * w.useIt();
        }

        return ammoPower * factor;
    }

    /**
     * @brief the shield is used, and it's protection is returned
     * @return the protection given by the shield
     * */
    @Override
    public float protection(){
        float factor = (float)1.0;  // This variable is used for acummulating the extra protection given by ShieldBoosters

        // The ShieldBoosters are used
        for(ShieldBooster s : shieldBoosters){
            factor = factor * s.useIt();
        }

        return shieldPower * factor;
    }

    /**
     * @brief makes the operations due to a received shot
     *        The shieldPower is reduced due  the the received shot
     * @param shot, the power of the received shot
     * @return the result of the received shot
     * */
    @Override
    public ShotResult receiveShot(float shot){
        if(protection() >= shot){
            // The shield is damaged
            shieldPower = shieldPower - SHIELDLOSSPERUNITSHOT * shot;   
            shieldPower = Math.max((float)0.0, shieldPower);

            // The space station resists the shot
            return ShotResult.RESIST;
        }else{
            // The shield is destroyed
            shieldPower = (float) 0.0;

            // The space station does not resist the shot
            return ShotResult.DONOTRESIST;
        }
    }


    /**
     * @brief Fuel units are decreased due to space station movement
     *        The decrement is the speed of the space station
     * */
    public void move(){
        fuelUnits = Math.max(0, fuelUnits * (1 - getSpeed()));
    }

    // Auxiliar Methods
    //==========================================================================
    /**
     * @brief Assigns fuel value without exceeding the limit
     * @param f, the new fuel value
     * */
    private void assignFuelValue(float f){
        fuelUnits = Math.min(f, MAXFUEL);
    }

    /**
     * @brief If pending damage has no effect, the pending damage is set to null
     * */
    private void cleanPendingDamage(){
        if(pendingDamage.hasNoEffect()){
            pendingDamage = null;
        }
    }
}
